File: gnd/src/main/java/com/google/android/gnd/ui/settings/SettingsFragment.java
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gnd.ui.settings;

import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.browser.customtabs.CustomTabColorSchemeParams;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.preference.Preference;
import androidx.preference.Preference.OnPreferenceChangeListener;
import androidx.preference.Preference.OnPreferenceClickListener;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;
import com.google.android.gnd.Config;
import com.google.android.gnd.R;

/**
 * Fragment containing app preferences saved as shared preferences.
 *
 * <p>NOTE: It uses {@link PreferenceFragmentCompat} instead of {@link
 * com.google.android.gnd.ui.common.AbstractFragment}, so dagger can't inject into it like it does
 * in other fragments.
 *
 * <p>TODO: Create dagger module and support injection into this fragment.
 */
public class SettingsFragment extends PreferenceFragmentCompat
    implements OnPreferenceChangeListener, OnPreferenceClickListener {

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    PreferenceManager preferenceManager = getPreferenceManager();
    preferenceManager.setSharedPreferencesName(Config.SHARED_PREFS_NAME);
    preferenceManager.setSharedPreferencesMode(Config.SHARED_PREFS_MODE);
  }

  @Override
  public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
    setPreferencesFromResource(R.xml.preferences, rootKey);
    for (String key : Keys.ALL_KEYS) {
      Preference preference = findPreference(key);
      if (preference == null) {
        throw new IllegalArgumentException("Key not found in preferences.xml: " + key);
      }

      preference.setOnPreferenceChangeListener(this);
      preference.setOnPreferenceClickListener(this);
    }
  }

  @Override
  public boolean onPreferenceChange(Preference preference, Object newValue) {
    switch (preference.getKey()) {
      case Keys.UPLOAD_MEDIA:
      case Keys.OFFLINE_AREAS:
        // do nothing.
        break;
      default:
        break;
    }
    return true;
  }

  @Override
  public boolean onPreferenceClick(Preference preference) {
    switch (preference.getKey()) {
      case Keys.VISIT_WEBSITE:
        openUrl(getString(R.string.ground_website));
        break;
      case Keys.FEEDBACK:
        openUrl(getString(R.string.ground_android_github));
        break;
      default:
        return false;
    }
    return true;
  }

  private void openUrl(String url) {
    CustomTabColorSchemeParams params =
        new CustomTabColorSchemeParams.Builder()
            .setToolbarColor(getResources().getColor(R.color.colorPrimary))
            .build();

    CustomTabsIntent customTabsIntent =
        new CustomTabsIntent.Builder().setDefaultColorSchemeParams(params).build();

    customTabsIntent.launchUrl(requireContext(), Uri.parse(url));
  }
}

File: app/src/main/java/org/groundplatform/android/ui/settings/SettingsFragment.kt
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.groundplatform.android.ui.settings

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.net.toUri
import androidx.core.os.LocaleListCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.preference.DropDownPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.SwitchPreferenceCompat
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.launch
import org.groundplatform.android.R
import org.groundplatform.android.common.Constants
import org.groundplatform.android.common.PrefKeys
import org.groundplatform.android.ui.common.ViewModelFactory
import org.groundplatform.android.ui.main.MainActivity

/** Fragment containing app preferences saved as shared preferences. */
@AndroidEntryPoint
class SettingsFragment : PreferenceFragmentCompat(), Preference.OnPreferenceClickListener {

  @Inject lateinit var viewModelFactory: ViewModelFactory
  private lateinit var viewModel: SettingsViewModel

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    viewModel = viewModelFactory[this, SettingsViewModel::class.java]
  }

  override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
    preferenceManager.sharedPreferencesName = Constants.SHARED_PREFS_NAME
    preferenceManager.sharedPreferencesMode = Constants.SHARED_PREFS_MODE

    setPreferencesFromResource(R.xml.preferences, rootKey)
    for (key in ALL_KEYS) {
      val preference =
        findPreference<Preference>(key)
          ?: throw IllegalArgumentException("Key not found in preferences.xml: $key")

      preference.onPreferenceClickListener = this
    }
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    viewLifecycleOwner.lifecycleScope.launch {
      viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
        viewModel.uiState.filterNotNull().collect { state ->
          val switchPreference = findPreference<SwitchPreferenceCompat>(PrefKeys.UPLOAD_MEDIA)
          switchPreference?.onPreferenceChangeListener =
            Preference.OnPreferenceChangeListener { _, _ ->
              viewModel.refreshUserPreferences()
              true
            }
          switchPreference?.isChecked = state.shouldUploadPhotosOnWifiOnly

          setupDropDownPreference(PrefKeys.LANGUAGE, state.language) { applyLocaleAndRestart(it) }
          setupDropDownPreference(PrefKeys.MEASUREMENT_UNITS, state.measurementUnits.name)
        }
      }
    }
  }

  private fun setupDropDownPreference(
    prefKey: String,
    selectedValue: String,
    onPrefChanged: (String) -> Unit = {},
  ) {
    findPreference<DropDownPreference>(prefKey)?.apply {
      updateSummary(selectedValue)
      onPreferenceChangeListener =
        Preference.OnPreferenceChangeListener { preference, newValue ->
          if (newValue is String) {
            updateSummary(newValue)
            onPrefChanged(newValue)
          }
          viewModel.refreshUserPreferences()
          true
        }
    }
  }

  private fun DropDownPreference.updateSummary(value: String) {
    val index = findIndexOfValue(value)
    if (index >= 0) {
      summary = entries[index]
    }
  }

  override fun onPreferenceClick(preference: Preference): Boolean {
    if (preference.key == PrefKeys.VISIT_WEBSITE) {
      openUrl(preference.summary.toString())
    }
    return true
  }

  private fun openUrl(url: String) {
    val intent = Intent(Intent.ACTION_VIEW, url.toUri())
    startActivity(intent)
  }

  private fun applyLocaleAndRestart(languageCode: String) {
    val appLocale = LocaleListCompat.forLanguageTags(languageCode)
    AppCompatDelegate.setApplicationLocales(appLocale)

    val intent =
      Intent(requireContext(), MainActivity::class.java).apply {
        flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
      }
    startActivity(intent)
  }

  companion object {
    private val ALL_KEYS =
      arrayOf(
        PrefKeys.LANGUAGE,
        PrefKeys.MEASUREMENT_UNITS,
        PrefKeys.UPLOAD_MEDIA,
        PrefKeys.VISIT_WEBSITE,
      )
  }
}
